### ISSUE TITLE
Update and re-instate FTPClient tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8053]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| amberite reported|
|-|-|
|Trac ID|trac#428|
|Type|defect|
|Created|2003-12-11 07:55:52Z|

Attachments:

* [test_ftp_428.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2ae/2aed8cef33a640e55d1a7f9358fc92db260a4de9/f491cd0eff0423c88ca52365b1c833a2b8270c29.diff) (7687 bytes) - added by therve on 2006-10-05 14:09:21Z - 
* [test_ftp_428.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2ae/2aed8cef33a640e55d1a7f9358fc92db260a4de9/dbba15016548e30697d84c63746f18cbecaf6bac.diff) (8280 bytes) - added by therve on 2006-10-06 10:37:29Z - 
* [test_ftp_428_3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2ae/2aed8cef33a640e55d1a7f9358fc92db260a4de9/5056b2d7c199c7e5a4e2af22b39e62a09fc23177.diff) (21504 bytes) - added by therve on 2006-10-10 10:26:16Z - 
* [test_ftp_428_4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2ae/2aed8cef33a640e55d1a7f9358fc92db260a4de9/f1f5b18f1c30138ebd977b070253a6107496c48e.diff) (22209 bytes) - added by therve on 2006-10-17 08:53:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__428 428
type__defect defect
reporter__amberite amberite
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1071129352000000 1071129352000000
changetime__1469610806112783 1469610806112783
version__ 
owner__ 
cc__spiv cc__slyphon cc__fdrake cc__amberite cc__exarkun cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|amberite commented|
|-|-|

```
#!html
<pre>
1.1.1 release of twisted contains this bug.  In ftp.py, FTPClient.storeFile breaks, because it relies on 
SenderProtocol, and SenderProtocol checks for error.ConnectionDone.  Since the error module gets 
imported from cred right after it gets imported from internet, checking for error.ConnectionDone 
raises an exception.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
And the FTPClient.storeFile unittest got removed in the server refactor, which
would have caught this! :(
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Here's a trivial (and untested) patch for this problem:

diff -u -r1.98 ftp.py
--- twisted/protocols/ftp.py    24 Nov 2003 15:18:09 -0000      1.98
+++ twisted/protocols/ftp.py    11 Dec 2003 02:10:53 -0000
@@ -32,7 +32,7 @@
 from math import floor
 
 # Twisted Imports
-from twisted.internet import abstract, reactor, protocol, error, defer
+from twisted.internet import abstract, reactor, protocol, defer
 from twisted.internet.interfaces import IProducer, IConsumer, IProtocol,
IFinishableConsumer
 from twisted.protocols import basic, policies
 from twisted.internet.protocol import ClientFactory, ServerFactory, Protocol, \
@@ -1561,6 +1561,7 @@
         self.connectedDeferred.callback(self)
         
     def connectionLost(self, reason):
+        from twisted.internet import error
         if reason.check(error.ConnectionDone):
             self.deferred.callback('connection done')
         else:
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="fdrake's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|fdrake commented|
|-|-|

```
#!html
<pre>
This is the same as #430.  The patch in that issue has been tested, though
it still doesn't have a unit test.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I've applied the patch from issue 430, which closes the original bug.

I'm leaving this issue open for the regression in unit testing that caused this
in the first place... the tests existed specifically to catch errors like this
that new code can introduce.  New functionality shouldn't remove tests, because
the tests are how you know the new functionality hasn't accidentally broken
anything else.  In cases like this where the APIs changed, then the tests should
be updated, not removed.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

I attached a patch adding a new testcase for all the methods in ftp.FTPClient. It also fixes and tests ftp.parsePWDResponse (which was not tested therefore broken :)).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

Updated the patch to work on windows.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

I made another update because test_STOR did not work. I used the patch in [#1794](https://github.com/twisted/twisted/issues/1794) to make the SenderProtocol works as expected.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

The test coverage re-added here is great.  If you don't have any more time to spend on this, I think it should probably be committed despite what I'm about to say, since it's far better than the previous state.

Since much of Twisted's test suite uses tests of the style included in this patch, I can understand why you wrote them this way.  However, there are a few problems with tests which set up real servers and connect to them with real clients.

  * They're prone to timeout issues
  * They're slower than tests which don't use actual network resources
  * They're slightly harder to debug when a failure is introduced (since they drag in a large amount of code for each unit being tested)
  * It's harder to see exactly what code is being covered, since a lot of code that ends up being run isn't explicitly part of the test, just part of the setup.

I would like you appeal to you to rewrite these tests without creating any sockets.  Instead, instantiate an FTPClient directly, connect it to a StringTransport (twisted.test.proto_helpers), feed it bytes manually with dataReceived and assert things about the bytes it writes to its transport.  Test the error paths as well as the positive paths (permission denied, out of disk space, etc).

If you could do this, I think the result would be FTP support in Twisted which is actual maintainable and approaching production quality.

What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

I agree on the complexity of the tests. They didn't feel unity at all: I actually stole the code that was previously in test_ftp. The main problem is wheter or not I'm able to write these tests using fake transports :). I'll give a try and send some feeback ASAP.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

OK I tried something. There are improvements possible (docs, factorization), but I wanted to know it's the way to go.

I set the review keyword to have some feedback.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

All the tests that involve Deferreds (eg, testCDUP) should return the Deferreds.  If the code is working properly, the assertion will get run synchronously and fail the test if the assertion fails, but it's possible the Deferred never fires, and the test will succeed without ever having asserted anything.

Good job hooking connectTCP.  The only thing I'd suggest is calling factory.buildProtocol in the test methods, instead of grabbing protoInstance directly (in case ftp ever decides it wants to change how its factory works).  ConnectionDone should be instantiated with 

String interpolations should be done on tuples, even for single-value interpolations.  So,

```
self.client.lineReceived('227 Entering Passive Mode (%s).' % (ftp.encodeHostPort('127.0.0.1', 12345),))
```

instead of what's there now.

One thing I don't quite understand about this version of the patch is the direct invocation of sendNextCommand that some of the tests do.  Why is this necessary?

Aside from the things I mentioned above, which are mainly quite minor, this patch is looking very good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_15):

Thanks a lot for your comments. 

> All the tests that involve Deferreds (eg, testCDUP) should return the Deferreds.  If the code is working properly, the assertion will get run synchronously and fail the test if the assertion fails, but it's possible the Deferred never fires, and the test will succeed without ever having asserted anything.

OK no problem.

> Good job hooking connectTCP.  The only thing I'd suggest is calling factory.buildProtocol in the test methods, instead of grabbing protoInstance directly (in case ftp ever decides it wants to change how its factory works). 

Right.

> ConnectionDone should be instantiated with 

Trac seems to have cut something here :).

> String interpolations should be done on tuples, even for single-value interpolations.

OK.

> One thing I don't quite understand about this version of the patch is the direct invocation of sendNextCommand that some of the tests do.  Why is this necessary?

For now that's the only way I found to queue the commands. As the 'lineReceived' happens in a callback, it doesn't let the client send the next command. I'll try to find a workaround.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

I attached an update, correcting most of the problems raised.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve removed owner|
|-|-|

Set for re-review. I think it mainly miss the correction around ConnectionDone.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Some thoughts:

  * Should FTPClient.pwd()'s Deferred fire with something which must be passed to parsePWDResponse?  A good, highish-level API would give back structured objects, not strings to be parsed.
  * Similar comment for FTPClient.cwd().
  * Are `pwd' and `cwd' really good names for these methods?

But those don't really have anything to do with this patch directly, since they are all existing problems.  The tests you've added look great.  I've put the patch into ftp-client-tests-428, expanded the docstrings and updated the test method names to the new naming convention we're using.  Someone else should take a look at this, since I've made these changes, but I think the branch is basically good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_20):
> Some thoughts:
>
>   * Should FTPClient.pwd()'s Deferred fire with something which must be passed to parsePWDResponse?  A good, highish-level API would give back structured objects, not strings to be parsed.
>   * Similar comment for FTPClient.cwd().
>   * Are `pwd' and `cwd' really good names for these methods?

Maybe we could add higher levels methods with differents names ? Like getDirectory/setDirectory, calling pwd/cwd and parsing the answers. I'll open a ticket.

> But those don't really have anything to do with this patch directly, since they are all existing problems.  The tests you've added look great.  I've put the patch into ftp-client-tests-428, expanded the docstrings and updated the test method names to the new naming convention we're using.  Someone else should take a look at this, since I've made these changes, but I think the branch is basically good to merge.

Great, I'll try to make some tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_26"></a>|@dreid set owner to @exarkun<br>@dreid set status to `new`|
|-|-|

Overall this is good.  However one little nit i have is the density of the testcases, I generally find it easier to read when there are newlines around defs and between asserts and the actions they are checking.  Pending successful completion of this branch on the buildbots you can merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [18892]) Merge ftp-client-tests-428

Author: therve, exarkun
Reviewer: exarkun, dreid
Fixes [#8053](https://github.com/twisted/twisted/issues/8053)

Add test coverage for twisted.protocols.ftp.FTPClient and fix several bugs
uncovered in the process.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_29"></a>|GitHub commented|
|-|-|

In changeset 0d4655736e709b449708e3c7d90c6a371dffa35f
```
#!CommitTicketReference repository="" revision="0d4655736e709b449708e3c7d90c6a371dffa35f"
Merge pull request #428 from twisted/8694-scop-osshcompat-py3-1

Author: scop
Reviewer: rodrigc
Fixes: #8694


Port twisted.conch.openssh_compat.* to Python 3
```
